### ISSUE TITLE
kotlin.x serialization: maps with enum keys are not supported

### ISSUE THREAD

**Issue by lamba92** [Litote/kmongo#196]
When trying to serialize [this](https://github.com/lamba92/dragalia-library/blob/dev/domain/src/commonMain/kotlin/com/github/lamba92/dragalialost/domain/entities/AdventurerEntity.kt) class a strange exception is thrown:

```text
writeString can only be called when State is VALUE, not when State is NAME
org.bson.BsonInvalidOperationException: writeString can only be called when State is VALUE, not when State is NAME
	at org.bson.AbstractBsonWriter.throwInvalidState(AbstractBsonWriter.java:746)
	at org.bson.AbstractBsonWriter.checkPreconditions(AbstractBsonWriter.java:701)
	at org.bson.AbstractBsonWriter.writeString(AbstractBsonWriter.java:606)
	at com.github.jershell.kbson.BsonEncoder.encodeEnum(BsonEncoder.kt:139)
	at kotlinx.serialization.internal.EnumSerializer.serialize(Enums.kt:70)
	at kotlinx.serialization.internal.EnumSerializer.serialize(Enums.kt:52)
	at kotlinx.serialization.Encoder$DefaultImpls.encodeSerializableValue(Encoding.kt:252)
	at kotlinx.serialization.builtins.AbstractEncoder.encodeSerializableValue(AbstractEncoder.kt:16)
	at kotlinx.serialization.builtins.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:72)
	at kotlinx.serialization.internal.MapLikeSerializer.serialize(CollectionSerializers.kt:131)
	at kotlinx.serialization.Encoder$DefaultImpls.encodeSerializableValue(Encoding.kt:252)
	at kotlinx.serialization.builtins.AbstractEncoder.encodeSerializableValue(AbstractEncoder.kt:16)
	at kotlinx.serialization.builtins.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:72)
	at ...
```
It seems that the internal state of the serializer is compromised when writing some elements. Also it happens only on enums.

Cross referenced on kbson [here](https://github.com/jershell/kbson/issues/15)

**Comment by zigzago** [comment]
Thanks for reporting. The fix is in the snapshot

**Comment by lamba92** [comment]
Could you push the snapshot on Sonatype snapshot repo? It is a blocking issue atm.

**Comment by zigzago** [comment]
@lamba92 the fix should be in the Sonatype snapshot (4.0.1-SNAPSHOT)
